### PR TITLE
[Feature] Support RTMDet-Ins and improve RTMDet test config.

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
1. Support RTMDet-Ins 
[PWC](https://paperswithcode.com/sota/real-time-instance-segmentation-on-mscoco?p=rtmdet-an-empirical-study-of-designing-real)

image

img_v2_d264acb8-5bfc-476a-a207-0c096f8da79g

3. Update RTMDet with YOLO series test config:
```
    test_cfg=dict(
        nms_pre=30000,
        min_bbox_size=0,
        score_thr=0.001,
        nms=dict(type='nms', iou_threshold=0.65),
        max_per_img=300),
```
For a fair comparison, we update the test_cfg in RTMDet.

## Modification

Add RTMDet-Ins
Update test_cfg
Support mask mosaic and mixup

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
4. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
5. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/transforms/transforms.py [hunk 1]
```diff
@@ -3532,16 +3554,32 @@ def transform(self, results: dict) -> dict:
         # 6. adjust bbox
         retrieve_gt_bboxes = retrieve_results['gt_bboxes']
         retrieve_gt_bboxes.rescale_([scale_ratio, scale_ratio])
+        if with_mask:
+            retrieve_gt_masks: BitmapMasks = retrieve_results[
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/transforms.py:3558]: It is not BitmapMasks.

File: mmdet/models/task_modules/assigners/dynamic_soft_label_assigner.py [hunk 2]
```diff
@@ -16,6 +16,17 @@
 EPS = 1.0e-7
 
 
+def center_of_mass(masks: Tensor, eps=1e-6):
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply mmdet/models/task_modules/assigners/dynamic_soft_label_assigner.py:19]: Provide a simple docstr

[the remark this answers is not in the document]
Comment (hhaAndroid) [reply mmdet/models/task_modules/assigners/dynamic_soft_label_assigner.py:19]: Provide a simple docstr

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 3]
```diff
@@ -0,0 +1,1019 @@
...
+from mmdet.utils import ConfigType, InstanceList, OptInstanceList, reduce_mean
+from .rtmdet_head import RTMDetHead
+
+
+@MODELS.register_module()
+class RTMDetInsHead(RTMDetHead):
+    """Detection Head of RTMDet-Ins.
+
+    Args:
+        num_prototypes (int): Number of mask prototype features extracted
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:31]: default values

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 4]
```diff
@@ -0,0 +1,1019 @@
...
+    def _init_layers(self) -> None:
+        """Initialize layers of the head."""
+        super()._init_layers()
+        self.kernel_convs = nn.ModuleList()
+        # calculate num dynamic parameters
+        weight_nums, bias_nums = [], []
+        for i in range(self.num_dyconvs):
+            if i == 0:
+                weight_nums.append(
+                    (self.num_prototypes + 2) * self.dyconv_channels)
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:68]: We should annotate the meaning of 2

File: configs/rtmdet/rtmdet-ins_l_8xb32-300e_coco.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        bbox_coder=dict(type='DistancePointBBoxCoder'),
+        loss_cls=dict(
+            type='QualityFocalLoss',
+            use_sigmoid=True,
+            beta=2.0,
+            loss_weight=1.0),
+        loss_bbox=dict(type='GIoULoss', loss_weight=2.0),
+        loss_mask=dict(
+            type='DiceLoss', loss_weight=2.0, eps=5e-6, reduction='mean')),
+    test_cfg=dict(
```
[the remark this answers is not in the document]
Comment (RangiLyu) [reply configs/rtmdet/rtmdet-ins_l_8xb32-300e_coco.py:25]: No, too much mask prediction will slow down the inference speed a lot.

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 6]
```diff
@@ -0,0 +1,1019 @@
...
+                bias_nums.append(self.dyconv_channels)
+            elif i == self.num_dyconvs - 1:
+                weight_nums.append(self.dyconv_channels)
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:71]:
```suggestion
                weight_nums.append(self.dyconv_channels * 1)
```

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 7]
```diff
@@ -0,0 +1,1019 @@
...
+            else:
+                weight_nums.append(self.dyconv_channels * self.dyconv_channels)
+                bias_nums.append(self.dyconv_channels)
+        self.weight_nums = weight_nums
+        self.bias_nums = bias_nums
+        self.num_gen_params = sum(weight_nums) + sum(bias_nums)
+
+        for i in range(self.stacked_convs):
+            chn = self.in_channels if i == 0 else self.feat_channels
+            self.kernel_convs.append(
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:82]: The concept of `kernel_convs` is not consistent with the original `controller,` so we suggest adding a comment

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 8]
```diff
@@ -0,0 +1,1019 @@
...
+        such as CenterNess in FCOS, IoU branch in ATSS.
+
+        Args:
+            cls_scores (list[Tensor]): Classification scores for all
+                scale levels, each is a 4D-tensor, has shape
+                (batch_size, num_priors * num_classes, H, W).
+            bbox_preds (list[Tensor]): Box energies / deltas for all
+                scale levels, each is a 4D-tensor, has shape
+                (batch_size, num_priors * 4, H, W).
+            score_factors (list[Tensor], optional): Score factor for
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply mmdet/models/dense_heads/rtmdet_ins_head.py:187]: docstring of kernel_preds

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 9]
```diff
@@ -0,0 +1,1019 @@
...
+
+        return self._bbox_post_process(
+            results=results,
+            mask_feat=mask_feat,
+            cfg=cfg,
+            rescale=rescale,
+            with_nms=with_nms,
+            img_meta=img_meta)
+
+    def _bbox_post_process(self,
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:411]:
```suggestion
    def _bbox_mask_post_process(self,
```

better?

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 10]
```diff
@@ -0,0 +1,1019 @@
...
+            h, w = img_meta['ori_shape'][:2] if rescale else img_meta[
+                'img_shape'][:2]
+            results.masks = torch.zeros(
+                size=(results.bboxes.shape[0], h, w),
+                dtype=torch.bool,
+                device=results.bboxes.device)
+
+        return results
+
+    def parse_dynamic_params(self, flatten_kernels):
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:506]: typehint

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 11]
```diff
@@ -0,0 +1,1019 @@
...
+                     batch_gt_instances_ignore: OptInstanceList = None):
+        """Compute losses of the head.
+
+        Args:
+            cls_scores (list[Tensor]): Box scores for each scale level
+                Has shape (N, num_anchors * num_classes, H, W)
+            bbox_preds (list[Tensor]): Decoded box for each scale
+                level with shape (N, num_anchors * 4, H, W) in
+                [tl_x, tl_y, br_x, br_y] format.
+            batch_gt_instances (list[:obj:`InstanceData`]): Batch of
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:662]: docstring of kernel_preds mask_feat

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 12]
```diff
@@ -0,0 +1,1019 @@
...
+        pos_gt_masks = torch.cat(pos_gt_masks, 0)
+        batch_pos_mask_logits = torch.cat(batch_pos_mask_logits, 0)
+
+        # avg_factor
+        num_pos = batch_pos_mask_logits.shape[0]
+        num_pos = reduce_mean(mask_feats.new_tensor([num_pos
+                                                     ])).clamp_(min=1).item()
+
+        if batch_pos_mask_logits.shape[0] == 0:
+            return mask_feats.sum() * 0, mask_feats.sum() * 0
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:623]:
```suggestion
            return mask_feats.sum() * 0
```

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 13]
```diff
@@ -0,0 +1,1019 @@
...
+            flatten_kernels (list[Tensor]): Kernels of the dynamic conv layers.
+                Has shape (N, num_instances, num_params)
+            sampling_results_list (list[:obj:`SamplingResults`]) Batch of
+                assignment results.
+            batch_gt_instances (list[:obj:`InstanceData`]): Batch of
+                gt_instance.  It usually includes ``bboxes`` and ``labels``
+                attributes.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:593]: Tensor

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 14]
```diff
@@ -0,0 +1,1019 @@
...
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        feat_channels: int = 256,
+        stacked_convs: int = 4,
+        num_levels: int = 3,
+        num_prototypes: int = 8,
+        act_cfg: ConfigType = dict(type='ReLU'),
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:767]:
```suggestion
        act_cfg: ConfigType = dict(type='ReLU', inplace=True),
```

File: mmdet/models/dense_heads/rtmdet_ins_head.py [hunk 15]
```diff
@@ -0,0 +1,1034 @@
...
+        for anchor, bbox_pred in zip(anchor_list[0], bbox_preds):
+            anchor = anchor.reshape(-1, 4)
+            bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(num_imgs, -1, 4)
+            bbox_pred = distance2bbox(anchor, bbox_pred)
+            decoded_bboxes.append(bbox_pred)
+
+        flatten_bboxes = torch.cat(decoded_bboxes, 1)
+        for gt_instances in batch_gt_instances:
+            gt_instances.masks = gt_instances.masks.to_tensor(
+                dtype=torch.bool, device=device)
```
Comment (tascj) [inline mmdet/models/dense_heads/rtmdet_ins_head.py:717]: The conversion overhead here is not negligible and it will slow down training. How about moving this to the data pipeline?

**Review by hhaAndroid — APPROVED** [review]

File: mmdet/models/dense_heads/rtmdet_ins_head.py (hunk 15 above)
Comment (RangiLyu) [reply mmdet/models/dense_heads/rtmdet_ins_head.py:717]: You are right. We are considering moving this to PackDetInputs. But this modification is related to all instance segmentation methods. We will open another PR for the refactoring.

**Review by zytx121 — DISMISSED** [review]
LGTM

File: README.md [hunk 16]
```diff
@@ -75,6 +75,24 @@ Apart from MMDetection, we also released [MMEngine](https://github.com/open-mmla
...
+[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/rtmdet-an-empirical-study-of-designing-real/object-detection-in-aerial-images-on-dota-1)](https://paperswithcode.com/sota/object-detection-in-aerial-images-on-dota-1?p=rtmdet-an-empirical-study-of-designing-real)
+[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/rtmdet-an-empirical-study-of-designing-real/object-detection-in-aerial-images-on-hrsc2016)](https://paperswithcode.com/sota/object-detection-in-aerial-images-on-hrsc2016?p=rtmdet-an-empirical-study-of-designing-real)
+
+| Task                     | Dataset | AP                                   | FPS(TRT FP16 BS1 3090) |
+| ------------------------ | ------- | ------------------------------------ | ---------------------- |
+| Object Detection         | COCO    | 52.8                                 | 322                    |
+| Instance Segmentation    | COCO    | 44.6                                 | 188                    |
```
Comment (tascj) [inline README.md:89]: @RangiLyu

Unlike Mask R-CNN which outputs`n x 28 x 28` RoI masks if we skip post processing, RTMDet-Ins outputs `n x img_h x img_w` by default. I think transferring those large masks from GPU to CPU will take quite a bit of time. Is the time included in latency/FPS measure?

Comment (RangiLyu) [reply README.md:89]: The decoding and rescaling of masks are included, but the GPU-CPU transferring is not. We tested the latency with `trtexec` and took the GPU Compute Time as the result (like the conventions in the previous YOLO series).

Comment (tascj) [reply README.md:89]: Thanks for the reply. It seems we need some modifications of the output masks for application.

Anyway, the mAP and GPU compute time is impressive. The model looks promising.

File: README.md [hunk 17]
```diff
@@ -75,6 +75,24 @@ Apart from MMDetection, we also released [MMEngine](https://github.com/open-mmla
 
 ## What's New
 
+### Highlight
+
+We are excited to announce our latest work on real-time object recognition tasks, **RTMDet**, a family of fully convolutional single-stage detectors. RTMDet not only achieves the best parameter-accuracy trade-off on object detection from tiny to extra-large model sizes, but also becomes the state-of-the-art on instance segmentation and rotated object detection tasks. Detailed explanation of its methodologies is available in our [technical report](https://arxiv.org/abs/2212.07784). Pre-trained models can be found [here](configs/rtmdet).
```
Comment (ZwwWayne) [inline README.md:80]:
```suggestion
We are excited to announce our latest work on real-time object recognition tasks, **RTMDet**, a family of fully convolutional single-stage detectors. RTMDet not only achieves the best parameter-accuracy trade-off on object detection from tiny to extra-large model sizes but also obtains new state-of-the-art performance on instance segmentation and rotated object detection tasks. Details can be found in the [technical report](https://arxiv.org/abs/2212.07784). Pre-trained models are [here](configs/rtmdet).
```

File: README.md [hunk 18]
```diff
@@ -75,6 +75,24 @@ Apart from MMDetection, we also released [MMEngine](https://github.com/open-mmla
```
Comment (ZwwWayne) [inline README.md:75]: Also update the section of `state of the art` section in major features.

Comment (ZwwWayne) [reply README.md:75]: The newly released [RTMDet](configs/rtmdet) also obtains new state-of-the-art results on real-time instance segmentation and rotated object detection tasks and the best parameter-accuracy trade-off on object detection.

File: configs/rtmdet/rtmdet_l_8xb32-300e_coco.py [hunk 19]
```diff
@@ -47,18 +47,18 @@
...
-        act_cfg=dict(type='SiLU')),
+        act_cfg=dict(type='SiLU', inplace=True)),
     train_cfg=dict(
         assigner=dict(type='DynamicSoftLabelAssigner', topk=13),
         allowed_border=-1,
         pos_weight=-1,
         debug=False),
     test_cfg=dict(
-        nms_pre=1000,
+        nms_pre=30000,
```
Comment (ZwwWayne) [inline configs/rtmdet/rtmdet_l_8xb32-300e_coco.py:57]: We can add comments in the config to indicate why and its outcome so that users can better understand it.
